### PR TITLE
txn: Use ResolveLockScan instead of ResolveLock & ResolveLockReadPhase

### PR DESCRIPTION
**Pull request by longfangsong**
Signed-off-by: longfangsong <longfangsong@icloud.com>

### What problem does this PR solve?

Problem Summary: 

Currently the `ResolveLock` & `ResolveLockReadPhase` will read out a batch of locks and resolve them together, which is hard to understand and unatural.

### What is changed and how it works?

What's Changed:
This pr propose a `ResolveLockScan` which "scan one lock and resolve one until the txn write size reach the limit" to replace the old `ResolveLock` & `ResolveLockReadPhase`

### Check List 

Tests 

- Unit test

Side effects

- Performance regression
    - maybe consumes more/less CPU
    - maybe consumes more/less MEM

**Additional context**

The behaviour of resolve locks in batches is introduced in #2389, which I don't really get why it is necessary.

### CODE REVIEW

File: src/storage/txn/commands/resolve_lock.rs [hunk 1]
```diff
@@ -1,135 +0,0 @@
...
-}
-
-impl CommandExt for ResolveLock {
-    ctx!();
-    tag!(resolve_lock);
-
-    command_method!(readonly, bool, false);
-    command_method!(is_sys_cmd, bool, true);
-
-    fn write_bytes(&self) -> usize {
```
Comment (longfangsong) [inline src/storage/txn/commands/resolve_lock.rs:55]: In fact, I think the old method of calculating write bytes might be wrong either.
if `txn.write_size() >= MAX_TXN_WRITE_SIZE` occurs, not every keys in `self.key_locks` will be written.

File: src/storage/txn/commands/resolve_lock_scan.rs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+impl CommandExt for ResolveLockScan {
+    ctx!();
+    tag!(resolve_lock);
+
+    command_method!(readonly, bool, false);
+    command_method!(is_sys_cmd, bool, true);
+
+    fn write_bytes(&self) -> usize {
+        // todo: how can I calculate this?
```
Comment (longfangsong) [inline src/storage/txn/commands/resolve_lock_scan.rs:54]: I think it is not possible to know `write_bytes` in advance if we take this approach.

File: src/storage/txn/scheduler.rs [hunk 3]
```diff
@@ -105,9 +105,9 @@ impl TaskContext {
         let tag = task.cmd.tag();
         let lock = task.cmd.gen_lock(latches);
         // Write command should acquire write lock.
-        if !task.cmd.readonly() && !lock.is_write_lock() {
-            panic!("write lock is expected for command {}", task.cmd);
-        }
+        // if !task.cmd.readonly() && !lock.is_write_lock() {
```
Comment (longfangsong) [inline src/storage/txn/scheduler.rs:108]: it is possibe for `ResolveLockScan` not generating any locks.

**Comment by longfangsong** [comment]
@nrc PTAL.
This is not finished yet, but I have already met up with some problems.

File: src/storage/txn/commands/resolve_lock_scan.rs [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        0
+    }
+
+    gen_lock!(first_scan_key: multiple); // it is not multiple actually, but Option is also IntoIterator
```
Comment (longfangsong) [inline src/storage/txn/commands/resolve_lock_scan.rs:58]: By doing this, we do not need to use callbacks for waiting on the latch. Just let the scheduler handle this.

**Comment by nrc** [comment]
@sticnarf PTAL

**Review by nrc — APPROVED** [review]

File: src/storage/txn/commands/resolve_lock_scan.rs [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+                        start_ts: current_lock.ts,
+                        commit_ts,
+                    }));
+                };
+                released_locks
+                    .entry(current_lock.ts)
+                    .or_insert_with(|| ReleasedLocks::new(current_lock.ts, commit_ts))
+                    .push(released);
+                rows += 1;
+                context.latches.release(&lock, context.cid);
```
Comment (sticnarf) [inline src/storage/txn/commands/resolve_lock_scan.rs:111]: The latch cannot be released until the mutation is applied. It should be released in `on_write_finished` which is the callback of `async_write`.

**Pull request closed without merging**